src/ssl.cpp

```diff
@@ -0,0 +1,257 @@
+		return;
+	}
+
+	// wrap the PEM certificate in a BIO, for openssl to read
+	BIO* bp = BIO_new_mem_buf(
+			const_cast<char*>(pem.data()),
+			int(pem.size()));
+
+	// parse the certificate into OpenSSL's internal representation
+	X509* cert = PEM_read_bio_X509_AUX(bp, nullptr, nullptr, nullptr);
+	BIO_free(bp);
+
+	if (!cert)
+	{
+		X509_STORE_free(cert_store);
+		ec = error_code(int(ERR_get_error()), error::get_ssl_category());
+		return;
+	}
+
+	// add cert to cert_store
+	X509_STORE_add_cert(cert_store, cert);
+	X509_free(cert);
+
+	// and lastly, replace the default cert store with ours
+	SSL_CTX_set_cert_store(c, cert_store);
+
+#elif defined TORRENT_USE_GNUTLS
+	std::string str(pem);
+    gnutls_datum_t ca;
+    ca.data = reinterpret_cast<unsigned char*>(const_cast<char*>(str.c_str())); // must be null terminated
+    ca.size = unsigned(str.size());
+
+	// Warning: returns the number of certificates processed or a negative error code on error
+	int ret = gnutls_certificate_set_x509_trust_mem(c.native_handle(), &ca, GNUTLS_X509_FMT_PEM);
+	if(ret < 0)
+		ec = error_code(ret, error::get_ssl_category());
+#endif
+}
+
+#if defined TORRENT_USE_OPENSSL
+namespace {
+
+int server_name_arg_ex_index = -1;
+
+int openssl_server_name_func(SSL* s, int*, void* arg)
+{
+	server_name_callback_type cb = reinterpret_cast<server_name_callback_type>(arg);
+	char const* name = SSL_get_servername(s, TLSEXT_NAMETYPE_host_name);
+	void* user_arg = SSL_CTX_get_ex_data(SSL_get_SSL_CTX(s), server_name_arg_ex_index);
+	return cb(s, name ? std::string(name) : "", user_arg)
+			? SSL_TLSEXT_ERR_OK
+			: SSL_TLSEXT_ERR_ALERT_FATAL;
+}
+
+} // anonymous namespace
+#endif
+
+void set_server_name_callback(context_handle_type c, server_name_callback_type cb, void* arg, error_code& ec)
+{
+#if defined TORRENT_USE_OPENSSL
+#if OPENSSL_VERSION_NUMBER >= 0x90812f
+	TORRENT_UNUSED(ec);
+#if defined __clang__
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wold-style-cast"
+#endif
+	if (server_name_arg_ex_index < 0)
+		server_name_arg_ex_index = SSL_get_ex_new_index(0, NULL, NULL, NULL, NULL);
+	SSL_CTX_set_ex_data(c, server_name_arg_ex_index, arg);
+
+	if(cb)
+	{
+		SSL_CTX_set_tlsext_servername_callback(c, openssl_server_name_func);
+		SSL_CTX_set_tlsext_servername_arg(c, reinterpret_cast<void*>(cb));
+	}
+	else {
+		SSL_CTX_set_tlsext_servername_callback(c, nullptr);
+		SSL_CTX_set_tlsext_servername_arg(c, nullptr);
+	}
+#if defined __clang__
+#pragma clang diagnostic pop
+#endif
+#else // OPENSSL_VERSION_NUMBER
+	ec = {boost::asio::error::operation_not_supported, error::get_ssl_category()};
+#endif
+
+#elif defined TORRENT_USE_GNUTLS
+	if(cb)
+		c.set_server_name_callback(
+				[cb, arg](stream_base& s, std::string const& name)
+				{
+					return cb(s, name, arg);
+				}
+				, ec);
+	else
+		c.set_server_name_callback(nullptr);
+#endif
+}
+
+void set_host_name(stream_handle_type s, std::string const& name, error_code& ec)
```

Comment:
Yes you are right!

The underlying issue is that GnuTLS does not have the same distinction of context and stream, so the native types exposed by `gnutls::context` and `gnutls::stream` can't be used as their OpenSSL counterpart. Especially for switching contexts, since the `gnutls::context` is actually only a credentials handler, the rest is emulated by the wrapper.